docker/main/ngen/py_funcs.py

```diff
@@ -0,0 +1,310 @@
+    parser.add_argument('--log-level', '-L', dest='log_level',
+                        default=os.environ.get("DEFAULT_LOG_LEVEL", "INFO").upper(),
+                        choices=["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"], help='Optionally specify log level.')
+
+    subparsers = parser.add_subparsers(dest='command', help='Specify the Python helper function command to run.')
+    subparsers.required = True
+
+    _parse_for_tar_and_copy(parent_subparsers_container=subparsers)
+    _parse_for_gather_output(parent_subparsers_container=subparsers)
+    _parse_for_move_job_output(parent_subparsers_container=subparsers)
+
+    return parser.parse_args()
+
+
+def _move_to_directory(source_dir: Path, dest_dir: Path, archive_name: Optional[str] = None):
+    """
+    Move source data files from their initial directory to a different directory, potentially combining into an archive.
+
+    Parameters
+    ----------
+    source_dir
+        The source directory containing output files to be placed within the dataset; note that this has already been
+        tested for existence as a directory.
+    dest_dir
+        The destination directory, to which the output data should be moved; note that this path has not yet been tested
+        for existence as a directory.
+    archive_name
+        When files should be archived as part of moving, the name of the archive; when files should not be archived,
+        ``None``.
+    """
+    if not dest_dir.is_dir():
+        raise ValueError(f"{get_date_str()} Can't move job output to non-directory path {dest_dir!s}!")
+
+    if archive_name:
+        logging.info("Archiving output files to output dataset")
+        tar_and_copy(source=source_dir, dest=dest_dir, archive_name=archive_name)
+    else:
+        logging.info("Moving output file(s) to output dataset")
+        for p in source_dir.glob("*"):
+            shutil.move(p, dest_dir)
+
+
+def gather_output(mpi_host_names: List[str], output_write_dir: Path):
+    """
+    Using subprocesses, gather output from remote MPI hosts and collect in the analogous directory on this host.
+
+    Parameters
+    ----------
+    mpi_host_names
+        List of MPI host names for the relevant job.
+    output_write_dir
+        Common job output directory across all hosts, from which data on remotes should be taken and in which data on
+        this host should be collected.
+    """
+    from socket import gethostname
+    local_hostname = gethostname()
+
+    scp_subs = OrderedDict()
+
+    for host in (h for h in mpi_host_names if h != local_hostname):
+        scp_subs[host] = Popen(f"scp -q -r {host}:${output_write_dir!s}/ ${output_write_dir!s}/.")
+
+    for host in scp_subs:
+        return_code = scp_subs[host].wait()
+        if return_code != 0:
+            raise RuntimeError(f"{get_date_str()} gather_output failed for {host} w/ return code {return_code!s}!")
```

Comment:
Try the following instead:

```python
for host, process in scp_subs.items():
    _, error_in_bytes = process.communicate()
    if process.returncode != 0:
        raise RuntimeError(f"{get_date_str()} gather_output failed for '{host}': {error_in_bytes.decode()} (code={process.returncode}!")
```

Reply:
I made a slight tweak to the error message for line formatting reasons but essentially adopted your suggestion.